neat_periodic_task/lib/neat_periodic_task.dart

```diff
@@ -0,0 +1,337 @@
+      delay = _timeout;
+    }
+
+    // If delay isn't past yet, we sleep.
+    if (elapsed < delay) {
+      Duration d = delay ~/ 2;
+      // Always sleep at least minCycle to ensure the iteration doesn't spin too
+      // fast as we approach the next iteration.
+      if (d < _minCycle) {
+        d = _minCycle;
+      }
+      // Never sleep more than maxCycle, as we must wake-up and print a log line
+      // that says we've checked the status of the task. Operators can either
+      // monitor the message saying this was done, or they can monitor the
+      // message saying that the status was monitored.
+      if (d > _maxCycle) {
+        d = _maxCycle;
+      }
+      if (status.state == 'finished') {
+        _log.info('### [ALIVE] neat-periodic-task: "$_name"');
+      }
+      _log.finest(() => 'NeatPeriodicTaskScheduler "$_name" sleeps $d');
+      await _sleep(d);
+
+      // Return such that we do another _iteration() call..
+      return;
+    }
+
+    // If elapsed >= delay, then we claim and run the task.
+    await _claimAndRun(status.update(
+      owner: Slugid.nice().toString(),
+      state: 'running',
+      updated: now,
+    ));
+  }
+
+  /// Claim the status and attempt to run.
+  Future<void> _claimAndRun(NeatTaskStatus status) async {
+    // Attempt to claim the task, if we're not the owner we failed to stop.
+    _log.finest(() => 'Attempt to claim task status for "$_name"');
+    final claimed = await _statusProvider.set(status.serialize());
+    if (!claimed) {
+      _log.finest(() => 'Failed to claim task status for "$_name"');
+      return;
+    }
+
+    try {
+      _log.info('### [START] neat-periodic-task: "$_name"');
+      await _task().timeout(_timeout);
+      _log.info('### [FINISHED] neat-periodic-task: "$_name"');
+    } on TimeoutException {
+      _log.shout(
+          '### [FAILED] neat-periodic-task: "$_name", timeout exceeded!');
+      return;
+    } catch (e, st) {
+      _log.shout('### [FAILED] neat-periodic-task: "$_name"', e, st);
+      return;
+    }
+
+    _log.finest(() => 'Attempting to set finished status for "$_name"');
+    final st = status.update(state: 'finished').serialize();
+    if (!await _statusProvider.set(st)) {
+      _log.info('Failed to set finished status for "$_name"');
+    }
+  }
+
+  /// Trigger the task, if it's not already running.
+  ///
+  /// This attempts to claim the status and trigger the task, if it is not
+  /// already running.
+  ///
+  /// Returns a [Future] that completes when the task is done.
+  Future<void> trigger() async {
+    _log.finest(() => 'NeatPeriodicTaskScheduler "$_name" trigger() called');
+    final status = await _getStatus();
+
+    // Find time elapsed since last status update.
+    final now = DateTime.now().toUtc();
+    final elapsed = now.difference(status.updated);
+
+    // If not running, or timed-out we run the task again.
+    if (status.state != 'running' || elapsed > _timeout) {
```

Comment:
Perhaps extract the status strings into constants (or enums).

Reply:
Yeah, in an ideal world... on the other hand it just adds a lot of typing for no particular value...